builtin/providers/aws/resource_aws_waf_rule.go

```diff
@@ -0,0 +1,202 @@
+									errors = append(errors, fmt.Errorf(
+										"%q must be one of IPMatch | ByteMatch | SqlInjectionMatch | SizeConstraint | XssMatch", k))
+								}
+								return
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsWafRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	// ChangeToken
+	var ct *waf.GetChangeTokenInput
+	time.Sleep(10 * time.Second) // To Prevent WAFStaleDataException
+	res, err := conn.GetChangeToken(ct)
+	if err != nil {
+		return fmt.Errorf("Error getting change token: %s", err)
+	}
+
+	params := &waf.CreateRuleInput{
+		ChangeToken: res.ChangeToken,
+		MetricName:  aws.String(d.Get("metric_name").(string)),
+		Name:        aws.String(d.Get("name").(string)),
+	}
+
+	resp, err := conn.CreateRule(params)
+	if err != nil {
+		return err
+	}
+	d.SetId(*resp.Rule.RuleId)
+	return resourceAwsWafRuleUpdate(d, meta)
+}
+
+func resourceAwsWafRuleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	params := &waf.GetRuleInput{
+		RuleId: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetRule(params)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "WAFNonexistentItemException" {
+			log.Printf("[WARN] WAF Rule (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	var predicates []map[string]interface{}
+
+	for _, predicateSet := range resp.Rule.Predicates {
+		predicate := map[string]interface{}{
+			"negated": *predicateSet.Negated,
+			"type":    *predicateSet.Type,
+			"data_id": *predicateSet.DataId,
+		}
+		predicates = append(predicates, predicate)
+	}
+
+	d.Set("predicates", predicates)
+	d.Set("name", resp.Rule.Name)
+	d.Set("metric_name", resp.Rule.MetricName)
+
+	return nil
+}
+
+func resourceAwsWafRuleUpdate(d *schema.ResourceData, meta interface{}) error {
+	err := updateWafRuleResource(d, meta, waf.ChangeActionInsert)
+	if err != nil {
+		return fmt.Errorf("Error Updating WAF Rule: %s", err)
+	}
+	return resourceAwsWafRuleRead(d, meta)
+}
+
+func resourceAwsWafRuleDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+	err := updateWafRuleResource(d, meta, waf.ChangeActionDelete)
+	if err != nil {
+		return fmt.Errorf("Error Removing WAF Rule Predicates: %s", err)
+	}
+	// ChangeToken
+	var ct *waf.GetChangeTokenInput
+	time.Sleep(5 * time.Second) // To Prevent WAFStaleDataException
+	resp, err := conn.GetChangeToken(ct)
+
+	req := &waf.DeleteRuleInput{
+		ChangeToken: resp.ChangeToken,
+		RuleId:      aws.String(d.Id()),
+	}
+	log.Printf("[INFO] Deleting WAF Rule")
+	_, err = conn.DeleteRule(req)
+
+	if err != nil {
+		return fmt.Errorf("Error deleting WAF Rule: %s", err)
+	}
+
+	return nil
+}
+
+func updateWafRuleResource(d *schema.ResourceData, meta interface{}, ChangeAction string) error {
```

Comment:
I think we need to check for change in the parameter before making a call to the API